### PR TITLE
Updated SR-IOV package

### PR DESCRIPTION
**Pull request by evgenLevin**

### CODE REVIEW

File: pkg/sriov/network.go [hunk 1]
```diff
@@ -353,6 +360,23 @@ func (builder *NetworkBuilder) withCapabilities(capability string) *NetworkBuild
 	return builder
 }
 
+func (builder *NetworkBuilder) withIpam(ipamType string) *NetworkBuilder {
```
Comment (trewest) [inline pkg/sriov/network.go:363]: can you move this method below all public methods/functions?

File: pkg/sriov/networknodestate.go [hunk 2]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
 
 	return true, nil
 }
+
+// WaitUntilSyncStatus waits for the duration of the defined timeout or until the
+// SriovNetworkNodeState gets to a specific syncStatus.
+func (builder *NetworkNodeStateBuilder) WaitUntilSyncStatus(syncStatus string, timeout time.Duration) error {
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:151]: can you add all public methods/functions above any private ones?

File: pkg/sriov/networknodestate.go [hunk 3]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+	glog.V(100).Infof("Waiting for the defined period until SriovNetworkNodeState %s has syncStatus %s",
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:152]: Please add a validate block at the top of this method

File: pkg/sriov/networknodestate.go [hunk 4]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+		builder.Objects.Name, syncStatus)
+
+	if builder.errorMsg != "" {
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:155]: This will be redundant once a validate block is added

File: pkg/sriov/networknodestate.go [hunk 5]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+		return fmt.Errorf(builder.errorMsg)
+	}
+
+	if syncStatus == "" {
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:159]: please add a glog entry for this

File: pkg/sriov/networknodestate.go [hunk 6]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+
+	// Polls every retryInterval to determine if SriovNetworkNodeState is in desired syncStatus.
+	return wait.PollImmediate(time.Second, timeout, func() (bool, error) {
+		err := builder.Discover()
+
+		if err != nil {
+			return false, nil
+		}
+
+		if builder.Objects.Status.SyncStatus == syncStatus {
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:171]: can be simplified to 
```suggestion
                return builder.Objects.Status.SyncStatus == syncStatus, nil
```

File: pkg/sriov/networknodestate.go [hunk 7]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+			return true, nil
+		}
+
+		return false, nil
+	})
+}
+
+// GetNumVFs returns num-vfs under the given interface.
+func (builder *NetworkNodeStateBuilder) GetNumVFs(sriovInterface string) (int, error) {
+	glog.V(100).Infof(
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:181]: please add validate block at top of this method

File: pkg/sriov/networknodestate.go [hunk 8]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+		"Getting num-vfs under interface %s from SriovNetworkNodeState %s",
+		sriovInterface, builder.nodeName)
+
+	if builder.Objects == nil {
+		return 0, fmt.Errorf(msg.UndefinedCrdObjectErrString("SriovNetworkNodeState"))
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:186]: please update errorMsg and return later

File: pkg/sriov/networknodestate.go [hunk 9]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+	}
+
+	if sriovInterface == "" {
+		glog.V(100).Infof("The sriovInterface  can not be empty string")
+
+		return 0, fmt.Errorf("the sriovInterface is empty sting")
```
Comment (trewest) [inline pkg/sriov/networknodestate.go:192]: please update errorMsg and return later

Comment (trewest) [reply pkg/sriov/networknodestate.go:192]: nit: is an empty string

File: pkg/sriov/networknodestatelist.go [hunk 10]
```diff
@@ -0,0 +1,62 @@
...
+
+		builder.errorMsg = "SriovNetworkNodeStateList 'nsname' is empty"
+	}
+
+	return builder
+}
+
+// Discover method gets the SriovNetworkNodeStateList items and stores them in the NetworkNodeStateListBuilder struct.
+func (builder *NetworkNodeStateListBuilder) Discover() error {
+	if builder.errorMsg != "" {
```
Comment (trewest) [inline pkg/sriov/networknodestatelist.go:50]: please add validate method and a validate block here

File: pkg/sriov/operatorConfig.go [hunk 11]
```diff
@@ -0,0 +1,89 @@
...
+	}
+
+	err := builder.apiClient.Update(context.TODO(), builder.Objects)
+	if err != nil {
+		return nil, fmt.Errorf("failed to update SriovOperatorConfig object %s in namespace %s",
+			builder.Objects.Name, builder.Objects.Namespace)
+	}
+
+	return builder, err
+}
```
Comment (trewest) [inline pkg/sriov/operatorConfig.go:89]: please add a validate method

File: pkg/sriov/operatorConfig.go [hunk 12]
```diff
@@ -0,0 +1,89 @@
...
+
+		builder.errorMsg = "SriovOperatorConfig 'nsname' is empty"
+	}
+
+	return builder
+}
+
+// Discover method gets the SriovOperatorConfig items and stores them in the OperatorConfigBuilder struct.
+func (builder *OperatorConfigBuilder) Discover() error {
+	if builder.errorMsg != "" {
```
Comment (trewest) [inline pkg/sriov/operatorConfig.go:59]: add validate block here

File: pkg/sriov/operatorConfig.go [hunk 13]
```diff
@@ -0,0 +1,89 @@
...
+	var err error
+	builder.Objects, err = builder.apiClient.SriovOperatorConfigs(builder.nsName).Get(
+		context.TODO(), builder.nodeName, v1.GetOptions{})
+
+	return err
+}
+
+// Update renovates the existing SriovOperatorConfig object with the SriovOperatorConfig definition in builder.
+func (builder *OperatorConfigBuilder) Update() (*OperatorConfigBuilder, error) {
+	glog.V(100).Infof("Updating the SriovOperatorConfig object %s in namespace %s",
```
Comment (trewest) [inline pkg/sriov/operatorConfig.go:75]: add validate block here

**Review by trewest — CHANGES_REQUESTED** [review]
Please address comments

File: pkg/sriov/networknodestate.go (hunk 2 above)
Comment (kononovn) [inline pkg/sriov/networknodestate.go:151]: I think you need to make sure that Sync is stable for few seconds before close this function. I suspect you can use this function as an example: https://github.com/openshift-kni/eco-goinfra/blob/b2486f52938008d0fa468e4e38a852ec60e9027f/pkg/mco/mcp.go#L248

File: pkg/sriov/networknodestate.go [hunk 14]
```diff
@@ -142,3 +145,62 @@ func (builder *NetworkNodeStateBuilder) validate() (bool, error) {
...
+		if builder.Objects.Status.SyncStatus == syncStatus {
+			return true, nil
+		}
+
+		return false, nil
+	})
+}
+
+// GetNumVFs returns num-vfs under the given interface.
+func (builder *NetworkNodeStateBuilder) GetNumVFs(sriovInterface string) (int, error) {
```
Comment (kononovn) [inline pkg/sriov/networknodestate.go:180]: Please change to sriovInterfaceName

File: pkg/sriov/networknodestatelist.go [hunk 15]
```diff
@@ -0,0 +1,62 @@
...
+	"github.com/golang/glog"
+
+	srIovV1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NetworkNodeStateListBuilder provides struct for SriovNetworkNodeStateList object
+// which contains connection to cluster and SriovNetworkNodeStateList definitions.
```
Comment (kononovn) [inline pkg/sriov/networknodestatelist.go:16]: SriovNetworkNodeStateList objects.

File: pkg/sriov/operatorConfig.go [hunk 16]
```diff
@@ -0,0 +1,89 @@
...
+	"github.com/golang/glog"
+
+	srIovV1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OperatorConfigBuilder provides struct for SriovOperatorConfigBuilder object which contains connection to cluster and
+// SriovOperatorConfig definitions.
```
Comment (kononovn) [inline pkg/sriov/operatorConfig.go:16]: SriovOperatorConfig objects.

File: pkg/sriov/operatorConfig.go [hunk 17]
```diff
@@ -0,0 +1,89 @@
...
+	// nodeName defines on what node SriovOperatorConfig resource should be queried.
+	nodeName string
+	// nsName defines SriovOperatorConfig operator namespace.
+	nsName string
+	// errorMsg used in discovery function before sending api request to cluster.
+	errorMsg string
+}
+
+// NewOperatorConfigBuilder creates new instance of SriovOperatorConfig.
+func NewOperatorConfigBuilder(apiClient *clients.Settings, nameConfig, nsname string) *OperatorConfigBuilder {
```
Comment (kononovn) [inline pkg/sriov/operatorConfig.go:31]: It looks line you are going to change this object.
Please remove discover func and NewOperatorConfigBuilder and replace it by PullOperatorConfigBuilder

File: pkg/sriov/operatorConfig.go [hunk 18]
```diff
@@ -0,0 +1,89 @@
...
+
+	srIovV1 "github.com/k8snetworkplumbingwg/sriov-network-operator/api/v1"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// OperatorConfigBuilder provides struct for SriovOperatorConfigBuilder object which contains connection to cluster and
+// SriovOperatorConfig definitions.
+type OperatorConfigBuilder struct {
```
Comment (kononovn) [inline pkg/sriov/operatorConfig.go:17]: If you are going to update this object add Defenition as well

**Review by kononovn — CHANGES_REQUESTED** [review]
Please check comments above

**Comment by kononovn** [comment]
Please update commit message with more info

File: pkg/sriov/operatorConfig.go (hunk 17 above)
Comment (evgenLevin) [reply pkg/sriov/operatorConfig.go]: structure removed as unnecessary

File: pkg/sriov/operatorConfig.go (hunk 18 above)
Comment (evgenLevin) [reply pkg/sriov/operatorConfig.go]: structure removed as unnecessary

File: pkg/sriov/operatorConfig.go (hunk 16 above)
Comment (evgenLevin) [reply pkg/sriov/operatorConfig.go]: structure removed as unnecessary

File: pkg/sriov/networknodestate.go (hunk 2 above)
Comment (evgenLevin) [reply pkg/sriov/networknodestate.go:125]: I think it's a little different. This function simply checks that the status is the same. There may be a scenario where the status will change after 5 seconds and this is expected

**Comment by evgenLevin** [comment]
@kononovn @trewest PTAL one more time. Thanks!

File: pkg/sriov/networknodestate.go (hunk 2 above)
Comment (kononovn) [reply pkg/sriov/networknodestate.go:125]: Following our conversation we need to implement multiple functions 
WaitUntilSyncStatus
WaitUntilInSyncStatusFor

File: pkg/sriov/networknodestate.go [hunk 19]
```diff
@@ -117,6 +120,67 @@ func (builder *NetworkNodeStateBuilder) GetNICs() (srIovV1.InterfaceExts, error)
...
+	glog.V(100).Infof(
+		"Getting num-vfs under interface %s from SriovNetworkNodeState %s",
+		sriovInterfaceName, builder.nodeName)
+
+	if builder.Objects == nil {
+		builder.errorMsg = msg.UndefinedCrdObjectErrString("SriovNetworkNodeState")
+	}
+
+	if sriovInterfaceName == "" {
+		glog.V(100).Infof("The sriovInterface  can not be empty string")
```
Comment (kononovn) [inline pkg/sriov/networknodestate.go:166]: Please remove 2 spaces

File: pkg/sriov/networknodestatelist.go [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+
+	v1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// NetworkNodeStateListBuilder provides struct for SriovNetworkNodeStateList objects
+// which contains connection to cluster and SriovNetworkNodeStateList definitions.
+type NetworkNodeStateListBuilder struct {
+	// Dynamically discovered SriovNetworkNodeStateList object.
+	Objects *srIovV1.SriovNetworkNodeStateList
```
Comment (kononovn) [inline pkg/sriov/networknodestatelist.go:19]: Please change it to []*NetworkNodeStateBuilder

File: pkg/sriov/policyList.go [hunk 21]
```diff
@@ -0,0 +1,62 @@
+package sriov
```
Comment (kononovn) [inline pkg/sriov/policyList.go:1]: I think it would be better to move this logic to list.go

File: pkg/sriov/networknodestatelist.go (hunk 20 above)
Comment (kononovn) [reply pkg/sriov/networknodestatelist.go:19]: 1 more option is to move this logic to list.go

**Review by kononovn — CHANGES_REQUESTED** [review]
looks good, just few more changes are required

File: pkg/sriov/list.go [hunk 22]
```diff
@@ -0,0 +1,80 @@
...
+	"context"
+	"fmt"
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ListNetworkNodeState returns NetworkNodeStates inventory in the given namespace.
+func ListNetworkNodeState(apiClient *clients.Settings,
```
Comment (kononovn) [inline pkg/sriov/list.go:13]: Please fix according to convention:
https://github.com/openshift-kni/eco-goinfra#functions-format

File: pkg/sriov/list.go [hunk 23]
```diff
@@ -0,0 +1,80 @@
...
+	nsname string, options metaV1.ListOptions) ([]*NetworkNodeStateBuilder, error) {
+	glog.V(100).Infof("Listing NetworkNodeStates in the nsname %s with the options %v", nsname, options)
```
Comment (kononovn) [inline pkg/sriov/list.go:15]: Change NetworkNodeStates to SriovNetworkNodeStates in all occurrences

File: pkg/sriov/network.go [hunk 24]
```diff
@@ -353,6 +360,23 @@ func (builder *NetworkBuilder) withCapabilities(capability string) *NetworkBuild
...
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
```
Comment (kononovn) [inline pkg/sriov/network.go:367]: Please check if ipamType is not empty string

**Review by kononovn — CHANGES_REQUESTED** [review]
Please fix few more comments

File: pkg/sriov/list.go (hunk 22 above)
Comment (evgenLevin) [reply pkg/sriov/list.go]: fixed

File: pkg/sriov/list.go (hunk 23 above)
Comment (evgenLevin) [reply pkg/sriov/list.go]: Changed

File: pkg/sriov/network.go (hunk 24 above)
Comment (evgenLevin) [reply pkg/sriov/network.go:367]: Added verification

**Review by kononovn — APPROVED** [review]
lgtm

File: pkg/sriov/network.go [hunk 25]
```diff
@@ -353,6 +360,29 @@ func (builder *NetworkBuilder) withCapabilities(capability string) *NetworkBuild
...
+		builder.errorMsg = "failed to configure IPAM, 'ipamType' parameter is empty"
+	}
+
+	builder.Definition.Spec.IPAM = fmt.Sprintf(`{ "type": "%s" }`, ipamType)
+
+	return builder
+}
+
+// GetSriovNetworksGVR returns SriovNetwork's GroupVersionResource which could be used for Clean function.
+func GetSriovNetworksGVR() schema.GroupVersionResource {
```
Comment (trewest) [inline pkg/sriov/network.go:380]: This is a public function. It should be above any private functions/methods

File: pkg/sriov/network.go [hunk 26]
```diff
@@ -353,6 +360,29 @@ func (builder *NetworkBuilder) withCapabilities(capability string) *NetworkBuild
...
+	if valid, _ := builder.validate(); !valid {
+		return builder
+	}
+
+	if ipamType == "" {
+		glog.V(100).Infof("sriov network 'ipamType' parameter can not be empty")
+
+		builder.errorMsg = "failed to configure IPAM, 'ipamType' parameter is empty"
+	}
+
```
Comment (trewest) [inline pkg/sriov/network.go:373]: please check if errorMsg != nil before updating struct

**Review by trewest — CHANGES_REQUESTED** [review]
Just a few comments

**Comment by evgenLevin** [comment]
@trewest PTAL. Thanks for your support!

File: pkg/sriov/list.go [hunk 27]
```diff
@@ -0,0 +1,80 @@
...
+
+	"github.com/golang/glog"
+	"github.com/openshift-kni/eco-goinfra/pkg/clients"
+	metaV1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// ListNetworkNodeState returns SriovNetworkNodeStates inventory in the given namespace.
+func ListNetworkNodeState(
+	apiClient *clients.Settings, nsname string, options metaV1.ListOptions) ([]*NetworkNodeStateBuilder, error) {
+	glog.V(100).Infof("Listing SriovNetworkNodeStates in the nsname %s with the options %v", nsname, options)
```
Comment (achuzhoy) [inline pkg/sriov/list.go:15]: nit: replace nsname with namespace

File: pkg/sriov/list.go [hunk 28]
```diff
@@ -0,0 +1,80 @@
...
+	if nsname == "" {
+		glog.V(100).Infof("SriovNetworkNodeStates 'nsname' parameter can not be empty")
+
+		return nil, fmt.Errorf("failed to list SriovNetworkNodeStates, 'nsname' parameter is empty")
+	}
+
+	networkNodeStateList, err := apiClient.SriovNetworkNodeStates(nsname).List(context.Background(), options)
+
+	if err != nil {
+		glog.V(100).Infof("Failed to list SriovNetworkNodeStates in the nsname %s due to %s", nsname, err.Error())
```
Comment (achuzhoy) [inline pkg/sriov/list.go:26]: ditto.

File: pkg/sriov/list.go [hunk 29]
```diff
@@ -0,0 +1,80 @@
...
+
+		networkNodeStateObjects = append(networkNodeStateObjects, stateBuilder)
+	}
+
+	return networkNodeStateObjects, nil
+}
+
+// ListPolicy returns SriovNetworkNodePolicies inventory in the given namespace.
+func ListPolicy(apiClient *clients.Settings, nsname string, options metaV1.ListOptions) ([]*PolicyBuilder, error) {
+	glog.V(100).Infof("Listing SriovNetworkNodePolicies in the nsname %s with the options %v",
```
Comment (achuzhoy) [inline pkg/sriov/list.go:49]: ditto

**Review by trewest — APPROVED** [review]
/lgtm - we can merge after addressing Sasha's comments

**Comment by evgenLevin** [comment]
All comments are fixed

**Pull request merged**